File: proveedor-viejo/orden_de_compra_service.py
from kafka.kafka_producer import KafkaProducer  
from datetime import datetime, timedelta
from sqlalchemy.orm import Session
from models import OrdenCompra, ItemOrdenCompra, OrdenDespacho, Producto

class OrdenDeCompraService:
    def __init__(self, db_session: Session):
        self.db_session = db_session
        self.kafka_producer = KafkaProducer()  

    def crear_orden_compra(self, tienda_id, estado, observaciones, items): 
        """
        Crea una nueva orden de compra y sus ítems asociados, y envía un mensaje a Kafka.
        Actualiza el stock de productos si la orden es aceptada.
        En caso de que no haya stock suficiente, la orden queda ACEPTADA, 
        pero se informa sobre los artículos faltantes.
        """
        errores, faltantes_stock = self.validar_items(items)
        
        estado, observaciones = self.generar_estado_observaciones(errores, faltantes_stock)
        print("Estado y observaciones generadas:", estado, observaciones)
        
        orden_compra = self.insertar_orden_compra(tienda_id, estado, observaciones)
        
        if not errores:
            self.insertar_items_orden_compra(orden_compra.id, items)

            if not faltantes_stock:
                self.actualizar_stock(items)
                orden_despacho_id = self.generar_orden_despacho(orden_compra.id)
                self.enviar_mensaje_despacho(tienda_id, orden_despacho_id, orden_compra.id)

        self.db_session.commit()

        self.enviar_mensaje_kafka(tienda_id, orden_compra.id, estado, observaciones, items)

        return orden_compra.id

    def validar_items(self, items):
        """Valida los ítems y devuelve errores y artículos faltantes de stock."""
        errores = []
        faltantes_stock = [] 
        productos_agrupados = {}

        for item in items:
            print(f"===> Procesando ítem: {item}")
            producto_id = item['producto_id']
            
            if item['cantidad'] < 1:
                error_msg = f"Artículo {producto_id}: cantidad mal informada."
                errores.append(error_msg)
                print(f"Error de cantidad: {error_msg}")
            
            if producto_id in productos_agrupados:
                productos_agrupados[producto_id]['cantidad'] += item['cantidad']
            else:
                productos_agrupados[producto_id] = {
                    'producto_id': producto_id,
                    'cantidad': item['cantidad']
                }

        print(f"Productos agrupados: {productos_agrupados}")
        
        for producto_id, datos in productos_agrupados.items():
            print(f"===> Validando stock del producto {producto_id} con cantidad total solicitada: {datos['cantidad']}")
            stock = self.db_session.query(Productos).filter(Productos.codigo == producto_id).first()

            print(f"Stock disponible para producto {producto_id}: {stock}")

            if not stock:
                error_msg = f"Artículo {producto_id}: no existe en el inventario."
                errores.append(error_msg)
                print(f"Error de inventario: {error_msg}")
            else:
                if datos['cantidad'] > stock.cantidad_stock_proveedor:
                    faltante_stock = {
                        'producto_id': producto_id,
                        'cantidad_solicitada': datos['cantidad'],
                        'cantidad_disponible': stock.cantidad_stock_proveedor
                    }
                    faltantes_stock.append(faltante_stock)
                    print(f"Stock insuficiente para {producto_id}: {faltante_stock}")

        print(f"Errores detectados: {errores}")
        print(f"Faltantes de stock: {faltantes_stock}")
        
        return errores, faltantes_stock

    def generar_estado_observaciones(self, errores, faltantes_stock):
        """Genera el estado y las observaciones de la orden."""
        estado = 'ACEPTADA'
        if errores:
            observaciones = '; '.join(errores)
            estado = 'RECHAZADA'
        elif faltantes_stock:
            observaciones = "Artículos faltantes: " + '; '.join([
                f"{falta['producto_id']}: solicitado {falta['cantidad_solicitada']}, disponible {falta['cantidad_disponible']}"
                for falta in faltantes_stock
            ])
            estado = 'PAUSADA'
        else:
            observaciones = ""

        return estado, observaciones

    def insertar_orden_compra(self, tienda_id, estado, observaciones):
        """Inserta la orden de compra en la base de datos y devuelve su objeto."""
        orden_compra = OrdenCompra(
            tienda_id=tienda_id,
            estado=estado,
            observaciones=observaciones
        )
        self.db_session.add(orden_compra)
        self.db_session.flush()  # Para obtener el ID después de agregar
        return orden_compra

    def insertar_items_orden_compra(self, orden_compra_id, items):
        """Inserta los ítems de la orden de compra en la base de datos."""
        for item in items:
            if item['cantidad'] > 0: 
                item_orden_compra = ItemOrdenCompra(
                    orden_compra_id=orden_compra_id,
                    producto_id=item['producto_id'],
                    color=item['color'],
                    talle=item['talle'],
                    cantidad=item['cantidad']
                )
                self.db_session.add(item_orden_compra)

    def actualizar_stock(self, items):
        """Actualiza el stock de productos en la base de datos."""
        for item in items:
            producto = self.db_session.query(Producto).filter(Producto.codigo == item['producto_id']).first()
            if producto:
                producto.cantidad_stock_proveedor -= item['cantidad']
                self.db_session.add(producto)

    def enviar_mensaje_kafka(self, tienda_id, orden_compra_id, estado, observaciones, items):
        """Envía un mensaje a Kafka con la información de la orden de compra."""
        mensaje_kafka = {
            'orden_id': orden_compra_id,
            'tienda_id': tienda_id,
            'estado': 'ACEPTADA' if estado == 'PAUSADA' else estado,
            'observaciones': observaciones,
            'items': items
        }
        self.kafka_producer.send_message(f"solicitudes", mensaje_kafka) 

    def obtener_ordenes_compra(self):
        """
        Retorna todas las órdenes de compra.
        """
        rows = self.db_session.query(OrdenCompra).all()
        print("Registros de órdenes de compra:", rows)  
        if not rows:
            print("No se encontraron órdenes de compra.")

        return [orden.__dict__ for orden in rows]

    def obtener_items_por_orden(self, orden_compra_id):
        """
        Retorna todos los ítems asociados a una orden de compra.
        """
        items = self.db_session.query(ItemOrdenCompra).filter(ItemOrdenCompra.orden_compra_id == orden_compra_id).all()
        return [item.__dict__ for item in items]

    def obtener_articulo_por_id(self, articulo_id):
        """
        Retorna un artículo por su ID.
        """
        articulo = self.db_session.query(Producto).filter(Producto.id == articulo_id).first()
        
        if articulo:
            return articulo.__dict__
        return None  
    
    def generar_orden_despacho(self, orden_compra_id):
        """Genera una nueva orden de despacho y devuelve su ID."""
        orden_despacho = OrdenDespacho(
            orden_compra_id=orden_compra_id,
            fecha_estimacion_envio=datetime.now() + timedelta(days=7),
            estado='PENDIENTE'
        )
        self.db_session.add(orden_despacho)
        self.db_session.flush()  # Para obtener el ID después de agregar
        return orden_despacho.id

    def enviar_mensaje_despacho(self, tienda_id, orden_despacho_id, orden_compra_id):
        """Envía un mensaje al topic de despacho con la información relevante."""
        fecha_estimacion_envio = datetime.now() + timedelta(days=7)
        
        mensaje_despacho = {
            'orden_despacho_id': orden_despacho_id,
            'orden_compra_id': orden_compra_id,
            'fecha_estimacion_envio': fecha_estimacion_envio.isoformat()  
        }
        
        self.kafka_producer.send_message(f"despacho", mensaje_despacho)  

    def marcar_orden_recibida(self, orden_id, despacho_id):
        """
        Marca una orden como recibida en la base de datos.
        """
        orden = self.db_session.query(OrdenCompra).filter(OrdenCompra.id == orden_id).first()

        if orden is None:
            print(f"Orden {orden_id} no existe en la base de datos.")
            return False

        estado = orden.estado 

        if estado != 'ACEPTADA':
            print(f"Orden {orden_id} no puede ser marcada como recibida porque no está en estado ACEPTADA. Está en {estado}")
            return False

        orden.fecha_recepcion = datetime.now()
        self.db_session.commit()

        print(f"Orden {orden_id} marcada como recibida con despacho {despacho_id}.")
        return True

File: proveedor/src/main/java/proveedor/service/OrdenDeCompraService.java
// OrdenDeCompraService.java
package proveedor.service;

import com.fasterxml.jackson.databind.ObjectMapper;

import proveedor.dto.ItemOrdenCompraDTO;
import proveedor.dto.OrdenCompraDTO;
import proveedor.service.KafkaProducerService;
import proveedor.models.*;
import proveedor.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class OrdenDeCompraService {

    @Autowired
    private OrdenCompraRepository ordenCompraRepository;

    @Autowired
    private ItemOrdenCompraRepository itemOrdenCompraRepository;

    @Autowired
    private ProductoRepository productoRepository;

    @Autowired
    private OrdenDespachoRepository ordenDespachoRepository;

    @Autowired
    private KafkaProducerService kafkaProducerService;

    @Autowired
    private TiendaRepository tiendaRepository;

    @Transactional
    public Long crearOrdenCompra(OrdenCompraDTO dto) {
        // Validar los ítems
        List<String> errores = new ArrayList<>();
        List<Map<String, Object>> faltantesStock = new ArrayList<>();

        Map<String, Integer> productosAgrupados = new HashMap<>();

        for (ItemOrdenCompraDTO item : dto.getItems()) {
            String productoId = item.getProducto_id();

            if (item.getCantidad() < 1) {
                errores.add("Artículo " + productoId + ": cantidad mal informada.");
                continue;
            }

            productosAgrupados.merge(productoId, item.getCantidad(), Integer::sum);
        }

        // Validar stock
        for (Map.Entry<String, Integer> entry : productosAgrupados.entrySet()) {
            String productoId = entry.getKey();
            Integer cantidadSolicitada = entry.getValue();

            Optional<Producto> productoOpt = productoRepository.findByCodigo(productoId);
            if (!productoOpt.isPresent()) {
                errores.add("Artículo " + productoId + ": no existe en el inventario.");
            } else {
                Producto producto = productoOpt.get();
                if (cantidadSolicitada > producto.getCantidadStockProveedor()) {
                    Map<String, Object> faltante = new HashMap<>();
                    faltante.put("producto_id", productoId);
                    faltante.put("cantidad_solicitada", cantidadSolicitada);
                    faltante.put("cantidad_disponible", producto.getCantidadStockProveedor());
                    faltantesStock.add(faltante);
                }
            }
        }

        // Generar estado y observaciones
        String estado;
        String observaciones;
        if (!errores.isEmpty()) {
            estado = "RECHAZADA";
            observaciones = String.join("; ", errores);
        } else if (!faltantesStock.isEmpty()) {
            estado = "PAUSADA";
            observaciones = "Artículos faltantes: " + faltantesStock.stream()
                    .map(f -> f.get("producto_id") + ": solicitado " + f.get("cantidad_solicitada") + ", disponible "
                            + f.get("cantidad_disponible"))
                    .collect(Collectors.joining("; "));
        } else {
            estado = "ACEPTADA";
            observaciones = "";
        }

        // Insertar orden de compra
        OrdenCompra ordenCompra = new OrdenCompra();
        // Suponiendo que tienes un TiendaRepository
        Optional<Tienda> tiendaOpt = tiendaRepository.findByCodigo(dto.getTienda_id());
        if (tiendaOpt.isPresent()) {
            ordenCompra.setTienda(tiendaOpt.get());
        } else {
            errores.add("Tienda " + dto.getTienda_id() + ": no existe.");
            // Manejar el error según corresponda
        }
        ordenCompra.setEstado(estado);
        ordenCompra.setObservaciones(observaciones);
        ordenCompra.setFechaSolicitud(OffsetDateTime.now());
        ordenCompra = ordenCompraRepository.save(ordenCompra);

        for (ItemOrdenCompraDTO itemDTO : dto.getItems()) {
            if (itemDTO.getCantidad() > 0) {
                ItemOrdenCompra itemOrdenCompra = new ItemOrdenCompra();
                itemOrdenCompra.setOrdenCompra(ordenCompra);

                // Recuperar el Producto existente
                Optional<Producto> productoOpt = productoRepository.findByCodigo(itemDTO.getProducto_id());
                if (productoOpt.isPresent()) {
                    itemOrdenCompra.setProducto(productoOpt.get());
                } else {
                    errores.add("Artículo " + itemDTO.getProducto_id() + ": no existe en el inventario.");
                    continue; // O manejar el error según corresponda
                }

                itemOrdenCompra.setColor(itemDTO.getColor());
                itemOrdenCompra.setTalle(itemDTO.getTalle());
                itemOrdenCompra.setCantidad(itemDTO.getCantidad());
                itemOrdenCompraRepository.save(itemOrdenCompra);
            }
        }

        // Actualizar stock si no hay errores ni faltantes
        if (errores.isEmpty() && faltantesStock.isEmpty()) {
            actualizarStock(dto.getItems());
            // Generar orden de despacho
            Long ordenDespachoId = generarOrdenDespacho(ordenCompra.getId());
            // Enviar mensaje de despacho
            enviarMensajeDespacho(dto.getTienda_id(), ordenDespachoId, ordenCompra.getId());
        }

        // Enviar mensaje a Kafka
        enviarMensajeKafka(dto.getTienda_id(), ordenCompra.getId(), estado, observaciones, dto.getItems());

        return ordenCompra.getId();
    }

    private void actualizarStock(List<ItemOrdenCompraDTO> items) {
        for (ItemOrdenCompraDTO item : items) {
            Optional<Producto> productoOpt = productoRepository.findByCodigo(item.getProducto_id());
            if (productoOpt.isPresent()) {
                Producto producto = productoOpt.get();
                producto.setCantidadStockProveedor(producto.getCantidadStockProveedor() - item.getCantidad());
                productoRepository.save(producto);
            }
        }
    }

    private Long generarOrdenDespacho(Long ordenCompraId) {
        OrdenDespacho ordenDespacho = new OrdenDespacho();
        OrdenCompra ordenCompra = ordenCompraRepository.findById(ordenCompraId)
                .orElseThrow(() -> new RuntimeException("OrdenCompra no encontrada"));
        ordenDespacho.setOrdenCompra(ordenCompra);
        ordenDespacho.setFechaEstimacionEnvio(OffsetDateTime.now().plusDays(7));
        ordenDespacho.setEstado("PENDIENTE");
        ordenDespacho = ordenDespachoRepository.save(ordenDespacho);
        return ordenDespacho.getId();
    }

    private void enviarMensajeDespacho(String tiendaId, Long ordenDespachoId, Long ordenCompraId) {
        Map<String, Object> mensajeDespacho = new HashMap<>();
        mensajeDespacho.put("orden_despacho_id", ordenDespachoId);
        mensajeDespacho.put("orden_compra_id", ordenCompraId);
        mensajeDespacho.put("fecha_estimacion_envio", OffsetDateTime.now().plusDays(7).toString());

        kafkaProducerService.sendMessage("despacho", mensajeDespacho);
    }

    private void enviarMensajeKafka(String tiendaId, Long ordenCompraId, String estado, String observaciones,
            List<ItemOrdenCompraDTO> items) {
        Map<String, Object> mensajeKafka = new HashMap<>();
        mensajeKafka.put("orden_id", ordenCompraId);
        mensajeKafka.put("tienda_id", tiendaId);
        mensajeKafka.put("estado", estado.equals("PAUSADA") ? "ACEPTADA" : estado);
        mensajeKafka.put("observaciones", observaciones);
        mensajeKafka.put("items", items);

        // Convertir el mensaje a JSON
        try {
            ObjectMapper mapper = new ObjectMapper();
            String mensajeJson = mapper.writeValueAsString(mensajeKafka);
            kafkaProducerService.sendMessage("solicitudes", mensajeJson);
        } catch (Exception e) {
            e.printStackTrace();
            // Manejar error de serialización
        }
    }

    public List<OrdenCompra> obtenerOrdenesCompra() {
        return ordenCompraRepository.findAll();
    }

    public List<ItemOrdenCompra> obtenerItemsPorOrden(Long ordenCompraId) {
        return itemOrdenCompraRepository.findByOrdenCompraId(ordenCompraId);
    }

    public Optional<Producto> obtenerArticuloPorId(Long articuloId) {
        return productoRepository.findById(articuloId);
    }

    @Transactional
    public boolean marcarOrdenRecibida(Long ordenId, Long despachoId) {
        Optional<OrdenCompra> ordenOpt = ordenCompraRepository.findById(ordenId);
        if (!ordenOpt.isPresent()) {
            System.out.println("Orden " + ordenId + " no existe en la base de datos.");
            return false;
        }

        OrdenCompra orden = ordenOpt.get();
        String estado = orden.getEstado();

        if (!estado.equals("ACEPTADA")) {
            System.out.println("Orden " + ordenId
                    + " no puede ser marcada como recibida porque no está en estado ACEPTADA. Está en " + estado);
            return false;
        }

        orden.setFechaRecepcion(OffsetDateTime.now());
        ordenCompraRepository.save(orden);

        System.out.println("Orden " + ordenId + " marcada como recibida con despacho " + despachoId + ".");
        return true;
    }

    @Transactional
    public boolean modificarEstadoOrdenCompra(Long ordenId) {
        OrdenCompra orden = ordenCompraRepository.findById(ordenId).orElse(null);
        if (orden == null) {
            return false;
        }

        List<String> errores = new ArrayList<>();
        List<String> faltantesStock = new ArrayList<>();

        List<ItemOrdenCompra> items = itemOrdenCompraRepository.findByOrdenCompraId(ordenId);

        for (ItemOrdenCompra item : items) {
            Producto producto = productoRepository.findById(item.getProducto().getId()).orElse(null);
            if (producto == null || item.getCantidad() < 1) {
                errores.add("Artículo " + item.getProducto().getId() + ": "
                        + (producto == null ? "no existe" : "cantidad mal informada"));
            } else if (item.getCantidad() > producto.getCantidadStockProveedor()) {
                faltantesStock.add(item.getProducto().getId().toString());
            }
        }

        String nuevoEstado;
        Map<String, Object> mensajeKafka = new HashMap<>();

        if (!errores.isEmpty()) {
            nuevoEstado = "RECHAZADA";
            String mensaje = String.join(", ", errores);
            mensajeKafka.put("orden_id", ordenId);
            mensajeKafka.put("nuevo_estado", nuevoEstado);
            mensajeKafka.put("errores", mensaje);
            kafkaProducerService.sendMessage("solicitudes", mensajeKafka);
        } else if (!faltantesStock.isEmpty()) {
            nuevoEstado = "PAUSADA";
            String mensaje = "Artículos sin stock: " + String.join(", ", faltantesStock);
            mensajeKafka.put("orden_id", ordenId);
            mensajeKafka.put("nuevo_estado", nuevoEstado);
            mensajeKafka.put("faltantes", mensaje);
            kafkaProducerService.sendMessage("solicitudes", mensajeKafka);
        } else {
            nuevoEstado = "ACEPTADA";
            mensajeKafka.put("orden_id", ordenId);
            mensajeKafka.put("nuevo_estado", nuevoEstado);
            kafkaProducerService.sendMessage("solicitudes", mensajeKafka);

            // Generar orden de despacho
            OrdenDespacho ordenDespacho = new OrdenDespacho();
            ordenDespacho.setOrdenCompra(orden);
            ordenDespacho.setFechaEstimacionEnvio(OffsetDateTime.now().plusDays(7));
            ordenDespacho.setEstado("PENDIENTE");
            ordenDespachoRepository.save(ordenDespacho);

            // Enviar mensaje de despacho
            Map<String, Object> mensajeDespacho = new HashMap<>();
            mensajeDespacho.put("orden_despacho_id", ordenDespacho.getId());
            mensajeDespacho.put("orden_compra_id", ordenId);
            mensajeDespacho.put("fecha_estimacion_envio", ordenDespacho.getFechaEstimacionEnvio().toString());

            kafkaProducerService.sendMessage("despacho", mensajeDespacho);

            // Restar stock
            for (ItemOrdenCompra item : items) {
                Producto producto = item.getProducto();
                producto.setCantidadStockProveedor(producto.getCantidadStockProveedor() - item.getCantidad());
                productoRepository.save(producto);
            }
        }

        // Actualizar estado de la orden
        orden.setEstado(nuevoEstado);
        ordenCompraRepository.save(orden);

        return true;
    }

}
